src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs

```diff
@@ -1894,5 +1882,467 @@ struct S
+            comp.VerifyEmitDiagnostics(
+                // (9,18): error CS8120: The switch case is unreachable. It has already been handled by a previous case or it is impossible to match.
+                //             case [..[var v]]: break;
+                Diagnostic(ErrorCode.ERR_SwitchCaseSubsumed, "[..[var v]]").WithLocation(9, 18));
+        }
+
+        [Fact]
+        public void Exhaustiveness_01()
+        {
+            var src = @"
+using System;
+class C
+{
+    public int X = 0, Y = 0;
+    public static void Test(Span<int> a, Span<C> b)
+    {
+        _ = a switch
+        {
+            [.., >=0] or [<0] or { Length: 0 or > 1 } => 0
+        };
+
+        _ = a switch
+        {
+            [.., >=0] or [..[.., <0] ] or [] => 0
+        };
+
+        _ = a switch
+        {
+            [..[>=0] ] or [<0] or
+            { Length: 0 or > 1 } => 0
+        };
+
+        _ = a switch
+        {
+            [..[.., <0]] or [..] => 0
+        };
+
+        _ = a switch
+        {
+            [_, ..{ Length: <= (int.MaxValue - 1) }] or [] => 0
+        };
+
+        _ = b switch
+        { 
+            [.., { X: >=0, Y: <0}] or [ { Y: >=0, X: <0} ] or
+            [.., { Y: >=0, X: <0}] or [ { X: >=0, Y: <0} ] or
+            [.., { X: <=0 }] or [{ X: >0 }] or 
+            { Length: 0 or > 1 } => 0
+        };
+    }
+}";
+            var comp = CreateCompilationWithIndexAndRangeAndSpan(src, parseOptions: TestOptions.RegularWithListPatterns);
+            comp.VerifyEmitDiagnostics();
+        }
+
+        [Fact]
+        public void Exhaustiveness_02()
+        {
+            var src = @"
+using System;
+class C
+{
+    public static void Test1(Span<int> a)
+    {
+        _ = a switch
+        {
+            [] => 1,
+            [_] => 2,
+            [_,..] => 3,
+        };
+    }
+    public static void Test2(Span<int> a)
+    {
+        _ = a switch
+        {
+            { Length: 0 } => 1,
+            { Length: 1 } => 2,
+            { Length: >=1 } => 3,
+        };
+    }
+}";
+            var comp = CreateCompilationWithIndexAndRangeAndSpan(src, parseOptions: TestOptions.RegularWithListPatterns);
+            comp.VerifyEmitDiagnostics();
+            var verifier = CompileAndVerify(comp);
+            string expectedIl = @"
+{
+  // Code size       26 (0x1a)
+  .maxstack  2
+  .locals init (int V_0)
+  IL_0000:  ldarga.s   V_0
+  IL_0002:  call       ""int System.Span<int>.Length.get""
+  IL_0007:  stloc.0
+  IL_0008:  ldloc.0
+  IL_0009:  brfalse.s  IL_0011
+  IL_000b:  ldloc.0
+  IL_000c:  ldc.i4.1
+  IL_000d:  beq.s      IL_0014
+  IL_000f:  br.s       IL_0017
+  IL_0011:  ldc.i4.1
+  IL_0012:  pop
+  IL_0013:  ret
+  IL_0014:  ldc.i4.2
+  IL_0015:  pop
+  IL_0016:  ret
+  IL_0017:  ldc.i4.3
+  IL_0018:  pop
+  IL_0019:  ret
+}
+";
+            verifier.VerifyIL("C.Test1", expectedIl);
+            verifier.VerifyIL("C.Test2", expectedIl);
+        }
```

Comment:
I'm wondering whether we should do something a little more subtle for possible values of Length. As this PR stands, we're breaking some existing scenarios (see example below).

I'm thinking we could do something like for nullability, whereby if you explicitly test for null on a non-nullable value then we take it seriously expand the value's domain (pure null tests).
The analog here would be that if you do a test for some negative value of Length, then we expand its value domain to include negative values.
I think that would avoid the breaking change.
This is not blocking this PR. Probably needs more reflection and discussion with LDM.

<break>
```
        [Fact]
        public void Subsumption_TODO2()
        {
            var src = @"
class C
{
    public int Length { get; set; }
    public int this[int i] => throw null;

    void Test()
    {
        _ = this switch
        {
            { Length: >= 0} => 0,
            { Length: -1 } => 1,
            _ => 2,
        };
    }
}";
            var comp = CreateCompilation(src);
            comp.VerifyEmitDiagnostics(
                // (12,13): error CS8510: The pattern is unreachable. It has already been handled by a previous arm of the switch expression or it is impossible to match.
                //             { Length: -1 } => 1,
                Diagnostic(ErrorCode.ERR_SwitchArmSubsumed, "{ Length: -1 }").WithLocation(12, 13)
                );
        }
````
</break>

Reply:
This was decided in https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-06-14.md
>We will specially recognize the `Count` and `Length` properties on types that are both countable and indexable, assuming that it can never be negative.

I see that it mentioned warnings but an impossible pattern or an no-exhaustive switch both are errors.

Reply:
At the time, `Length` and `Count` meant the implicit references due to list-pattern (which is new syntax so not breaking). 
But since then we've expanded it to also mean direct checks like `{ Length: ... }` so we'll need to discuss break.

Reply:
Correction: non-exhaustive switches give warnings, but I'm not sure if an inconsistency was intended. I'll leave it as-is until confirmed.

Reply:
I added a note to test plan to follow up with LDM on this part of the breaking change.
This doesn't block this PR.